### DISCUSSION TITLE
How to check if a tree sequence has any locally unary nodes

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tskit#2865] [Show and tell]
We want to do this in tsdate (see https://github.com/tskit-dev/tsdate/issues/320). Here's the fastest way I have found (takes about 90 secs on my laptop for a 138MB / 258547-tree ts ([chr 2q](https://zenodo.org/record/5512994/files/hgdp_tgp_sgdp_high_cov_ancients_chr2_q.dated.trees.tsz) from the unified genealogy):

```python
def has_locally_unary_nodes(ts):
    for tree, ediff in zip(ts.trees(), ts.edge_diffs()):
        changed = {e.parent for edges in (ediff.edges_out, ediff.edges_in) for e in edges}
        if (tree.num_children_array[list(changed)] == 1).any():
            return True
    return False
```

@jeromekelleher  mentioned in that tskit issue that there might be a speedier way using numba, though.

**Comment by jeromekelleher** [comment]
I'm surprised messing with the edge diffs makes any difference - out of interest what's the time difference with:
```python
for tree in ts.trees():
    if np.any(tree.num_children_array == 1):
           return True
return False
```

But yeah, this can definitely be done more quickly using the numba incremental algorithm approach.

**Comment by barneyhill** [comment]
Yeah double-checked this on the quebec ts (1050794 ts, 564.8 MiB) I average 22 seconds on the has_locally_unary_nodes and 3 minutes with Jerome's method.
